src/messages.rs

```diff
@@ -0,0 +1,6 @@
+use crate::{error::PeerNetResult, peer_id::PeerId};
+
+pub trait MessagesHandler: Clone + Send + 'static {
+    /// Get a message from a byte array received from the network deserialize it and handle it
+    fn deserialize_and_handle(&self, data: &[u8], peer_id: &PeerId) -> PeerNetResult<()>;
```

Comment:
We could use associated types to get something more generic here, something like:

``` rust
pub trait MessageHandler {
   type Message;
   fn deserialize(&self, data: &[u8]) -> PeerNetResult<Self::Message>;
   fn handle(&self, peer_id: &PeerId, msg: &Self::Message) -> PeerNetResult<()>;
}
```
It's not much more complicated, and would allow peeking a message without handling it

Reply:
The problem is that in common cases like the one in Massa we want to deserialize only the id in the current thread and then send the rest to the handler to deserialize and handle the message. With you logic it implies deserializing the whole message before sending it to the handler.

You can see the current implementation in the PR in massa

Reply:
Well then I think it wouldn't be too ugly to have a `fn deserialize_id(&self, data: &[u8]) -> PeerNetResult<(usize, PeerId)>;`
(returning the offset for later read)
But that's just a personal test, the more "atomic" functions, the better we can test and adapt it to our needs

Reply:
Yes ok I could split them.

Reply:
Done